kernel/src/device/evdev/mod.rs

```diff
@@ -0,0 +1,365 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Event device (evdev) support.
+//!
+//! Character device with major number 13. The minor numbers are dynamically allocated.
+//! Devices appear as `/dev/input/eventX` where X is the minor number.
+//!
+//! Reference: <https://elixir.bootlin.com/linux/v6.17/source/include/uapi/linux/major.h>
+
+mod file;
+
+use alloc::{
+    format,
+    string::String,
+    sync::{Arc, Weak},
+    vec::Vec,
+};
+use core::{
+    fmt::Debug,
+    sync::atomic::{AtomicU32, Ordering},
+    time::Duration,
+};
+
+use aster_input::{
+    event_type_codes::{EventTypes, SynEvent},
+    input_dev::{InputDevice, InputEvent},
+    input_handler::{ConnectError, InputHandler, InputHandlerClass},
+};
+use aster_time::read_monotonic_time;
+use device_id::{DeviceId, MajorId, MinorId};
+use file::{EvdevEvent, EvdevFile, EVDEV_BUFFER_SIZE};
+use ostd::sync::SpinLock;
+use spin::Once;
+
+use super::char::{acquire_major, register, unregister, CharDevice, MajorIdOwner};
+use crate::{
+    device::char::DevtmpfsName,
+    fs::inode_handle::FileIo,
+    prelude::*,
+    syscall::ClockId,
+    time::clocks::{
+        BootTimeClock, MonotonicClock, MonotonicCoarseClock, MonotonicRawClock, RealTimeClock,
+        RealTimeCoarseClock,
+    },
+    util::ring_buffer::RbProducer,
+};
+
+/// Major device number for evdev devices.
+const EVDEV_MAJOR_ID: u16 = 13;
+
+/// Global minor number allocator for evdev devices.
+static EVDEV_MINOR_COUNTER: AtomicU32 = AtomicU32::new(0);
+
+/// Global registry of evdev devices.
+static EVDEV_DEVICES: SpinLock<Vec<(u32, Arc<EvdevDevice>)>> = SpinLock::new(Vec::new());
+
+pub struct EvdevDevice {
+    /// Minor device number.
+    minor: u32,
+    /// Reference count of open clients.
+    open: AtomicU32,
+    /// Input device associated with this evdev.
+    device: Arc<dyn InputDevice>,
+    /// List of clients with their producers.
+    ///
+    /// # Deadlock Prevention
+    ///
+    /// This lock is acquired in both the task and interrupt contexts.
+    /// We must make sure that this lock is taken with the local IRQs disabled.
+    /// Otherwise, we would be vulnerable to deadlock.
+    client_list: SpinLock<Vec<(Weak<EvdevFile>, RbProducer<EvdevEvent>)>>,
```

Comment:
Again, the concept of "client" no longer exists. Rename all references to "clients" as "opened evdev files". For example, this field should be named to "opened_files" or "open_table".